### DISCUSSION TITLE
Header access API v2

### DISCUSSION THREAD

**Discussion by bagder** [curl/curl#8496] [Ideas]
Hello again,

After discussions and lots of feedback on the first API proposal, here comes a much updated [second take on an API designed for HTTP header access](https://github.com/curl/curl/wiki/get-headers-v2).

What do you miss? Would it work? How can it be further improved?

**Comment by icing** [comment]
Reviewed it this morning. It contains all one needs to access the original information that was sent from the server, I think.

Comments on documentation:

* the struct is tied to the `easy` handle, but reused on subsequent calls, I understand. So,  different easy handles will all have their own and no one steps on each others toes?
* When *exactly* will which headers/trailers be available/complete? Seems that trailers will only be there when the content is read, obviously. Maybe an example  of reading a response and when headers are supposed to be inspected would help?
* Are callers to copy the key/values as well or when will they go out of scope?
* There is no way to specify that one is looking for a header/trailer value specifically, right? Will headers of same name always come before trailers?


I find the exposed struct to work, but exposing many details. Also that it is an `easy-static` is ok, but...hm..

An alternative would be a `do()` callback function that visits all headers /trailers. Something like this:

```
typedef int curl_header_cb(
  CURL *easy, 
  const char *name, 
  const char *val, 
  int index, 
  int flags, 
  void *userdata);

curl_easy_header_do(
  CURL *easy,  /* handle to inspect */
  const char *name,  /* name of field or NULL for all */
  int flags,  /* 0 for all, 1 for HEADERS, 2 for TRAILERS, 3 -> 0 */
  curl_header_cb *cb, 
  void *userdata);

```

But mostly, this is a matter of taste.

**Reply by bagder** [reply]
> the struct is tied to the `easy` handle, but reused on subsequent calls, I understand. So,  different easy handles will all have their own and no one steps on each others toes?

Correct. Since the  `easy` handle is tied and associated with a single transfer so it also work fine when doing multiple transfers using the multi interface.

> When _exactly_ will which headers/trailers be available/complete? Seems that trailers will only be there when the content is read, obviously. Maybe an example  of reading a response and when headers are supposed to be inspected would help?

Yeah, we can work on clarifying this. My thinking was to allow this API to be used at any time, including from all callbacks, so headers would be available from when they have (completely) arrived even before the entire set has arrived.

But if asking for details before the response is complete, maybe we should set a flag in the struct saying so? Since an additional header received later might then actually change things for a receiver. Like if you ask for a cookie header, you won't know how many there are until all headers have arrived.

> Are callers to copy the key/values as well or when will they go out of scope?

My thinking is that callers should copy the values the want to keep. They go out of scope on the next header-call using the same easy handle - and when the easy handle is freed with `curl_easy_cleanup()`.

> There is no way to specify that one is looking for a header/trailer value specifically, right?

Oh right. Maybe that's something one might want so that users don't accidentally mix them up.

I suppose we need to add a `flags` argument to `curl_easy_header` for that, so that a caller can ask for one, the other or both...

> Will headers of same name always come before trailers?

I think we can define that, yes. And the trailer will be identified by the flags bit in the struct.

> I find the exposed struct to work, but exposing many details. Also that it is an `easy-static` is ok, but...hm..

Agreed. But I tried a few different variations in my head before I landed on this being the best I could think of...

**Comment by bagder** [comment]
One big thing struck me: If there's a redirect followed, I've just presumed that we start over and only record and provide headers from the most recent set of headers. I need to clarify that in the docs.

**Comment by MonkeybreadSoftware** [comment]
Why not "nextHeader" pointer in struct curl_header  to the next header in the list?
And also could have a "nextValue" pointer to the next struct curl_header  for the same header with following value?
Then we could traverse all headers with a while() loop until nextHeader is NULL.
And optionally look into the values for multiple headers with a second inner while loop.

```
struct curl_header {
    struct curl_header * nextHeader; /* next name */
    struct curl_header * nextValue; /* same name, more values */
   char *name;    /* this might not use the same case */
   char *value;
   size_t amount; /* number of headers using this name  */
   size_t index;  /* ... of this instance, 0 or higher */
   unsigned int flags; /* extra info */
   void *private; /* handle privately used by libcurl */
};
```

This would skip calling curl_easy_nextheader a lot of times.


You could also define that calling curl_easy_header with name = NULL returns first header.


Also all new structures should have reserved pointer to leave room for future expansion of the structures with ABI compatibility.

**Reply by bagder** [reply]
> This would skip calling curl_easy_nextheader a lot of times.

Most applications would iterate over the headers somehow anyway so I don't think having to call `curl_easy_nextheader` multiple times would be too problematic.

The price is basically having to build that entire tree at once (and we probably need to provide a function that can free it). If we would do such a function, I would change the struct a little since `amount` for example then wouldn't have to be in all nodes and the `next*` pointers wouldn't be used when `curl_easy_header()` is used.

> all new structures should have reserved pointer to leave room for future expansion of the structures with ABI compatibility.

I'm hoping we can come up with a struct that is "complete enough" already when we remove the EXPERIMENTAL tag and thus we won't have to expand it. But the `private` pointer in there also gives us room to do all sorts of fun, should we need to.

**Comment by xquery** [comment]
not to expand scope - though was curious how https://datatracker.ietf.org/doc/html/rfc8941 might be relevant here and if we thought such an api should provide parity in helping create header slist (CURLOPT_HTTPHEADER) ?

**Reply by bagder** [reply]
**Structured Field Values** is exactly that: a defined standard syntax for header values. This API helps applications get those values but it doesn't parse them. Applications will still need to parse the values to understand them.

**Comment by bagder** [comment]
Since there was no particular objections or problems identified, we have polished the API slightly after feedback and this has now gotten quiet again, my conclusion is that this is now a decent proposal. The next step is now to write up a first version to allow us to test it and get a feel for how it would work out in the real world.

**Comment by bagder** [comment]
It struck me that we haven't discussed how this API should or shouldn't work in regards to CONNECT response headers when going through a proxy for HTTPS to the remote server. Should we just ignore those headers and leave them out from this API, or should we provide them as well somehow?

**Comment by bagder** [comment]
Work in progress: #8593

**Comment by jzakrzewski** [comment]
@bagder Sooner or later somebody will come very surprised "how could you miss such an important and obvious thing?", so I'd say to tackle it right away, so that it won't be "frankensteined" on otherwise nice API. Maybe the end result would be the same but at least it would come from a design process and not because the API would be set in stone, and the only way to add such functionality would be X.

**Reply by bagder** [reply]
Agreed. I've now made the CONNECT headers accessible using the same API. The user just needs to specify that with a bit in the function call.

**Comment by dfandrich** [comment]
* The proposal is pretty HTTP-centric right now. How would other protocols be handled? Many of them have a similar concept to HTTP headers that could be useful to expose, like FTP or RTSP, or even proxy protocols like HAPROXY or SOCKS (where the authentication negotiation could be made available this way).
* How would the "headers" be handled in ping-pong-like protocols (if at all)?
* How about the case of chained proxies, like SOCKS→HTTP→HTTPS or  SOCKS→HTTP→FTP?
* What happens in the case of HTTP redirects? Which headers are returned then?

I'm not saying that all these need to be implemented or implemented right away, just that they should be considered so that it's possible to add them cleanly later if there's at all a chance that they'd be wanted in the future, e.g. instead of one bit for a proxy maybe 8 bits to specify a proxy "level", or maybe a way to specify the protocol that particular headers belong to.

**Comment by bagder** [comment]
> The proposal is pretty HTTP-centric right now. How would other protocols be handled? Many of them have a similar concept to HTTP headers that could be useful to expose, like FTP or RTSP, or even proxy protocols like HAPROXY or SOCKS (where the authentication negotiation could be made available this way).

It is indeed very HTTP-centric (and documented so far to be for HTTP-only). I decided to focus on that to make sure it doesn't get made too generic at the expense of making it less good for HTTP.

While there are similar concepts in other protocols, I've not got the feeling for nor the requests from users for this kind of functionality for those protocols.

But I'm open for ideas what what we can/should do to avoid "closing doors" or how this API can be expanded (or not) to other protocols.

> What happens in the case of HTTP redirects? Which headers are returned then?

Right now, this API is only for the most recent request in a redirect (or even multi-pass auth) "sequence". Mostly because I haven't come up with a decent way to expose the response headers from multiple requests in a convenient and practical way.

I'll make sure the documentation states this clearly.

> instead of one bit for a proxy maybe 8 bits to specify a proxy "level", or maybe a way to specify the protocol that particular headers belong to.

While working on the PR for this I switched the single bit over to a numeric "type" for where the header originates from (I now also updated the wiki to show my current API take) as I couldn't really make the bit work with the different header origins we have.

**Comment by bagder** [comment]
I think I'll switch the `type` in the `curl_easy_header()` back to a bitmask so that a user can ask for a header from one **or more** of the header origins. Also, I'll add the same mask to the `curl_easy_nextheader()` function so that you can opt to iterate over the right subset of headers.

**Reply by bagder** [reply]
I've updated [the wiki page](https://github.com/curl/curl/wiki/get-headers-v2) to match my current PR version.

**Comment by icing** [comment]
Looking at the  current design, I like the `origin` flag. I find the term `SERVER` a bit unclear. In httpd, I use the term  `final` response for the real answer to the request, in contrast to `intermediate` responses.

What I would also like is to iterate the headers by response: iterate all headers of response 0, 1, 2...or just iterate over the `final` headers (-1?). But then, one would also like to see the status codes of the ith response and then the API design of "just" headers does not fit any longer.

I wonder if iterating responses (and their headers) are what people really want to do with this API?

**Reply by bagder** [reply]
>  I find the term SERVER a bit unclear. In httpd, I use the term final response for the real answer to the request, in contrast to intermediate responses.

I'm not terribly happy with it either and would love to improve it. I started out with "NORMAL" but that didn't really work for me either. Ideally it should be something that separates it from "TRAILER" decently. I'm not sure I think "final" to be much better. Maybe "HEADER" is the best we can do...?

If the API should provide headers from multiple responses, like when a redirect or multi-stage authentication is used, then "final" could be used to refer to the final response. But then of course there can be *both* headers and trailers in all responses.

**Reply by icing** [reply]
Ah, forgot the redirect cases. Hmm. Then, to make it semantically correct, exposing responses as entities becomes really helpful. And `final` becomes really the `last` response.

pseudo coding (excuse the misaligns):
```
#define CURLH_HEADER       (1 << 0)
#define CURLH_TRAILER       (1 << 1)
#define CURLH_PSEUDO       (1 << 2)  /* this is a pseudo header like ':status' */

#define CURLH_FIRST           (0)
#define CURLH_FINAL           (-1)
#define CURLH_NTH(n)        (n)

struct curl_header *curl_easy_startheader(CURL *easy,
                                         unsigned int response_index,
                                         unsigned int htype);
```

(updated to remove last param)

**Reply by bagder** [reply]
I think you're right and that I should extend the API able to provide headers from the entire "chain" of responses that the transfer involved. Maybe we should reverse the order though so that last is 0, as that's the response that is most likely to be interesting for users and then we go backwards by increasing, or decreasing, the counter.

**Reply by icing** [reply]
It should increase. Because then you can give access to responses before the final one has arrived.

**Reply by bagder** [reply]
Ah right, as otherwise the numbers change as new responses arrive...

**Reply by bagder** [reply]
And that point about psuedo headers was good point too. I don't have them added yet...

**Reply by icing** [reply]
You could also add a `meta` (does one still dare use that word) type where you expose other, internal info. Timing would be one idea.

**Reply by bagder** [reply]
Hm. I'm not too thrilled about that. I think it would complicate the function a little too much.

**Reply by icing** [reply]
Agreed, just pointing out the mere possibility. 😌

**Reply by bagder** [reply]
I updated the wiki page accordingly. Now both functions also take a `request` argument to specify which request to get the headers from. 0 being the first and -1 being a shortcut to the latest.

**Reply by icing** [reply]
Small nit: I think a `curl_easy_startheader()` is a better design and saves you from checking if response/origin are still the same on a `nextheader` call. And what are the semantics of that anyway?

**Reply by bagder** [reply]
Right, you mean having two functions for this instead of overloading a single one. I'll think about it.

**Reply by bagder** [reply]
The function doesn't actually check if they are the same in subsequent calls, you *can* change them - but that will only make the returned set very odd.

**Reply by icing** [reply]
If you keep it with one function - which is fine - the documentation should state the when `prev != NULL`, `origin` and `request` are ignored and keep their effective value from the fist call. (if that is how you intend it to work, of course)

**Reply by bagder** [reply]
I think I like the single function approach better because otherwise we would have to store the origin + request values somehow from the first invoke and use for the second function, which adds more logic.

The current man page for `curl_easy_nextheader()`:

NAME
====

curl\_easy\_nextheader - get the next HTTP header

SYNOPSIS
========

    #include <curl/curl.h>

    struct curl_header *curl_easy_nextheader(CURL *easy,
                                             unsigned int origin,
                                             int request,
                                             struct curl_header *prev);

DESCRIPTION
===========

EXPERIMENTAL feature!

This function lets an application iterate over all previously received
HTTP headers.

The *origin* argument is for specifying which headers to receive, as a
single HTTP transfer might provide headers from several different places
and they may then have different importance to the user and headers
using the same name might be used. The *origin* is a bitmask for what
header sources you want. See the *curl\_easy\_header(3)* man page for
the origin descriptions.

The *request* argument tells libcurl from which request you want headers
from. A single transfer might consist of a series of HTTP requests and
this argument lets you specify which particular invidual request you
want the headers from. 0 being the first request and then the number
increases for further redirects or when multi-state authentication is
used. Passing in -1 is a shortcut to \"the last\" request in the series,
independently of the actual amount of requests used.

It is suggested that you pass in the same **origin** and **request**
when iterating over a range of headers as changing the value mid-loop
might give you unexpected results.

If *prev* is NULL, this function returns a pointer to the first header
stored within the given scope (origin + request).

If *prev* is a pointer to a previously returned header struct,
*curl\_easy\_nextheader(3)* returns a pointer the next header stored
within the given scope. This way, an application can iterate over all
availble headers.

The memory for the struct this points to, is owned and managed by
libcurl and is associated with the easy handle. Applications must copy
the data if they want it to survive subsequent API calls or the
life-time of the easy handle.

EXAMPLE
=======

    struct curl_header *prev = NULL;
    struct curl_header *h;

    /* extract the normal headers from the first request */
    while((h = curl_easy_nextheader(easy, CURLH_HEADER, 0, prev))) {
       print "%s: %s\n", h->name, h->value);
       prev = h;
    }

    /* extract the normal headers + 1xx + trailers from the last request */
    unsigned int origin = CURLH_HEADER| CURLH_1XX | CURLH_TRAILER;
    while((h = curl_easy_nextheader(easy, origin, -1, prev))) {
       print "%s: %s\n", h->name, h->value);
       prev = h;
    }

AVAILABILITY
============

Added in 7.83.0

RETURN VALUE
============

This function returns the next header, or NULL when there are no more
(matching) headers or an error occurred.

If this function returns NULL when *prev* was set to NULL, then there
are no headers available within the scope to return.

SEE ALSO
========

**curl\_easy\_header**(3), **curl\_easy\_perform**(3)

**Comment by bagder** [comment]
# HTTP response header access in curl

This support depends on the new headers API work coming in [PR 8593](https://github.com/curl/curl/pull/8593). Hopefully landing in 7.83.0.

This output is done with `-w` and new magic added to its abilties.

## Output individual header contents

Say hello to `%header{name}` where name is the case insensitive name of the header withon colon.

~~~
$ curl -w '%header{date}\n' curl.se
Thu, 17 Mar 2022 16:59:24 GMT
~~~

## Output all response headers as JSON

Say hello to `%{header_json}`

~~~
$ curl -sw '%{header_json}\n' curl.se | jq
{
  "Server": "Varnish",
  "Retry-After": "0",
  "Location": "https://curl.se/",
  "Content-Length": "0",
  "Accept-Ranges": "bytes",
  "Date": "Thu, 17 Mar 2022 17:01:38 GMT",
  "Via": "1.1 varnish",
  "Connection": "close",
  "X-Served-By": "cache-bma1667-BMA",
  "X-Cache": "HIT",
  "X-Cache-Hits": "0",
  "X-Timer": "S1647536498.479515,VS0,VE0"
}
~~~

**Comment by bagder** [comment]
PR #8593 for the new headers API has now been merged!

I blogged about it: https://daniel.haxx.se/blog/2022/03/22/a-headers-api-for-libcurl/
